File: server/routes/labels.js
import express from 'express'
import { readLabels, updateLabels, readAutoLabels, updateAutoLabels } from '../lib/labelStore.js'
import { VALID_LABELS } from '../lib/validLabels.js'
import { isAuthed } from './auth.js'

const router = express.Router()

const VALID = VALID_LABELS

// GET /api/labels/valid — return the canonical label list
router.get('/valid', (_req, res) => res.json({ labels: VALID_LABELS }))

// GET /api/labels — return all labels (PRIVATE: full map requires login).
// `labels` holds human-confirmed labels only (the training set); `suggestions`
// holds agent-predicted labels awaiting review in Label Studio.
router.get('/', async (req, res) => {
  if (!isAuthed(req)) return res.status(401).json({ error: 'Login required' })
  try {
    const labels      = await readLabels()
    const suggestions = await readAutoLabels()
    res.json({ labels, suggestions })
  } catch (err) {
    res.status(500).json({ error: 'Failed to read labels', detail: err.message })
  }
})

// POST /api/labels/:filename/suggest — record the agent's predicted label as a
// review suggestion. Deliberately NOT written to labels.json: the model's own
// predictions must never feed back into its training data unreviewed.
router.post('/:filename/suggest', async (req, res) => {
  const { filename } = req.params
  const { label } = req.body

  if (!filename.startsWith('recording-') || !filename.endsWith('.webm')) {
    return res.status(400).json({ error: 'Invalid filename' })
  }
  if (!VALID.includes(label)) {
    return res.status(400).json({ error: `Invalid label: ${label}` })
  }

  try {
    await updateAutoLabels(auto => { auto[filename] = label; return auto })
    res.json({ filename, suggested: label })
  } catch (err) {
    res.status(500).json({ error: 'Failed to save suggestion', detail: err.message })
  }
})

// POST /api/labels/:filename — set label for a clip
router.post('/:filename', async (req, res) => {
  const { filename } = req.params
  const { label } = req.body

  if (!filename.startsWith('recording-') || !filename.endsWith('.webm')) {
    return res.status(400).json({ error: 'Invalid filename' })
  }
  if (!VALID.includes(label)) {
    return res.status(400).json({ error: `Invalid label: ${label}` })
  }

  try {
    await updateLabels(labels => { labels[filename] = label; return labels })
    // The human has reviewed this clip — the agent's suggestion is resolved.
    await updateAutoLabels(auto => { delete auto[filename]; return auto }).catch(() => {})
    res.json({ filename, label })
  } catch (err) {
    res.status(500).json({ error: 'Failed to save label', detail: err.message })
  }
})

// DELETE /api/labels/:filename — remove label (and any suggestion) from a clip
router.delete('/:filename', async (req, res) => {
  const { filename } = req.params
  try {
    await updateLabels(labels => { delete labels[filename]; return labels })
    await updateAutoLabels(auto => { delete auto[filename]; return auto }).catch(() => {})
    res.json({ deleted: filename })
  } catch (err) {
    res.status(500).json({ error: 'Failed to delete label', detail: err.message })
  }
})

export default router

File: server_py/app/routers/labels.py
"""Port of server/routes/labels.js."""
from fastapi import APIRouter, Request
from fastapi.responses import JSONResponse
from pydantic import BaseModel

from app.auth import is_authed
from app.lib.label_store import read_auto_labels, read_labels, update_auto_labels, update_labels
from app.lib.recording_name import is_recording_filename
from app.lib.valid_labels import VALID_LABELS

router = APIRouter()


class LabelBody(BaseModel):
    label: str | None = None


@router.get("/valid")
async def get_valid_labels():
    return {"labels": VALID_LABELS}


@router.get("")
@router.get("/")
async def get_labels(request: Request):
    """PRIVATE: full map requires login. `labels` holds human-confirmed labels
    only (the training set); `suggestions` holds agent-predicted labels
    awaiting review in Label Studio."""
    if not is_authed(request):
        return JSONResponse(status_code=401, content={"error": "Login required"})
    try:
        labels = await read_labels()
        suggestions = await read_auto_labels()
        return {"labels": labels, "suggestions": suggestions}
    except Exception as err:
        return JSONResponse(status_code=500, content={"error": "Failed to read labels", "detail": str(err)})


@router.post("/{filename:path}/suggest")
async def suggest_label(filename: str, body: LabelBody):
    """Record the agent's predicted label as a review suggestion. Deliberately
    NOT written to labels.json: the model's own predictions must never feed
    back into its training data unreviewed."""
    if not is_recording_filename(filename):
        return JSONResponse(status_code=400, content={"error": "Invalid filename"})
    if body.label not in VALID_LABELS:
        return JSONResponse(status_code=400, content={"error": f"Invalid label: {body.label}"})
    try:
        await update_auto_labels(lambda auto: {**auto, filename: body.label})
        return {"filename": filename, "suggested": body.label}
    except Exception as err:
        return JSONResponse(status_code=500, content={"error": "Failed to save suggestion", "detail": str(err)})


@router.post("/{filename:path}")
async def set_label(filename: str, body: LabelBody):
    """Set label for a clip."""
    if not is_recording_filename(filename):
        return JSONResponse(status_code=400, content={"error": "Invalid filename"})
    if body.label not in VALID_LABELS:
        return JSONResponse(status_code=400, content={"error": f"Invalid label: {body.label}"})
    try:
        await update_labels(lambda labels: {**labels, filename: body.label})
        # The human has reviewed this clip — the agent's suggestion is resolved.
        try:
            await update_auto_labels(lambda auto: {k: v for k, v in auto.items() if k != filename})
        except Exception:
            pass
        return {"filename": filename, "label": body.label}
    except Exception as err:
        return JSONResponse(status_code=500, content={"error": "Failed to save label", "detail": str(err)})


@router.delete("/{filename:path}")
async def delete_label(filename: str):
    """Remove label (and any suggestion) from a clip."""
    if not is_recording_filename(filename):
        return JSONResponse(status_code=400, content={"error": "Invalid filename"})
    try:
        await update_labels(lambda labels: {k: v for k, v in labels.items() if k != filename})
        try:
            await update_auto_labels(lambda auto: {k: v for k, v in auto.items() if k != filename})
        except Exception:
            pass
        return {"deleted": filename}
    except Exception as err:
        return JSONResponse(status_code=500, content={"error": "Failed to delete label", "detail": str(err)})
